### PR TITLE
Add eviction/removal listeners

### PR DESCRIPTION
**Pull request by matteocastellotti**
Added the possibility to add an eviction/removal listener when configuring the cache

### CODE REVIEW

**Comment by alvarosanchez** [comment]
@matteocastellotti thanks for your contribution! Could you please add a note at https://github.com/micronaut-projects/micronaut-cache/blob/master/src/main/docs/guide/cache-abstraction.adoc mentioning that this is possible? Otherwise, this would go unnoticed.

**Comment by graemerocher** [comment]
Thanks for considering contributing, but I am not sure this implementation makes sense as written:

1. It works only for caffeine even though the configuration is added to cache-core
2. It has no tests whatsoever

**Review by ludovicc — CHANGES_REQUESTED** [review]

**Comment by matteocastellotti** [comment]
I added a new configuration layer for the parameters related to caffeine cache and the test to verify that the implementation works

**Comment by alvarosanchez** [comment]
There are some Checkstyle failures

File: cache-caffeine/src/main/java/io/micronaut/cache/caffeine/configuration/DefaultCacheConfiguration.java [hunk 1]
```diff
@@ -27,7 +27,8 @@
  * @since 1.0.2
  */
 @EachProperty(CacheConfiguration.PREFIX)
-public class DefaultCacheConfiguration extends CacheConfiguration {
+public class DefaultCacheConfiguration extends CaffeineCacheConfiguration {
```
Comment (graemerocher) [inline cache-caffeine/src/main/java/io/micronaut/cache/caffeine/configuration/DefaultCacheConfiguration.java:30]: this class is public, moving it another package breaks binary compatibility. To accept this change we would have to bump the major version. Alternatively you could consider not moving this class to a different package.

File: cache-caffeine/src/main/java/io/micronaut/cache/caffeine/DefaultSyncCache.java [hunk 2]
```diff
@@ -189,7 +187,7 @@ public void put(@NonNull Object key, @Nullable Object value) {
      * @param cacheConfiguration The cache configurations
      * @return cache
      */
-    protected com.github.benmanes.caffeine.cache.Cache buildCache(CacheConfiguration cacheConfiguration) {
+    protected com.github.benmanes.caffeine.cache.Cache buildCache(CaffeineCacheConfiguration cacheConfiguration) {
```
Comment (graemerocher) [inline cache-caffeine/src/main/java/io/micronaut/cache/caffeine/DefaultSyncCache.java:190]: This breaks binary compatibility, you would need to revert this change and introduce an instanceof check

File: cache-caffeine/src/main/java/io/micronaut/cache/caffeine/DefaultSyncCache.java [hunk 3]
```diff
@@ -81,7 +79,7 @@ public DefaultSyncCache(
      */
     @Inject
     public DefaultSyncCache(
-            CacheConfiguration cacheConfiguration,
+            CaffeineCacheConfiguration cacheConfiguration,
```
Comment (graemerocher) [inline cache-caffeine/src/main/java/io/micronaut/cache/caffeine/DefaultSyncCache.java:82]: This breaks binary compatibility, you would need to revert this change and introduce an instanceof check

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: cache-caffeine/src/main/java/io/micronaut/cache/caffeine/configuration/DefaultCacheConfiguration.java (hunk 1 above)
Comment (matteocastellotti) [reply cache-caffeine/src/main/java/io/micronaut/cache/caffeine/configuration/DefaultCacheConfiguration.java]: Done

File: cache-caffeine/src/main/java/io/micronaut/cache/caffeine/DefaultSyncCache.java (hunk 2 above)
Comment (matteocastellotti) [reply cache-caffeine/src/main/java/io/micronaut/cache/caffeine/DefaultSyncCache.java]: Done

File: cache-caffeine/src/main/java/io/micronaut/cache/caffeine/DefaultSyncCache.java (hunk 3 above)
Comment (matteocastellotti) [reply cache-caffeine/src/main/java/io/micronaut/cache/caffeine/DefaultSyncCache.java]: Done

**Review by graemerocher — APPROVED** [review]

**Comment by graemerocher** [comment]
Thanks for the contribution!
